File: src/air_mouse_cpp/src/mission_manager.cpp
#include <rclcpp/rclcpp.hpp>
#include <nav_msgs/msg/occupancy_grid.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <std_msgs/msg/string.hpp>
#include <std_msgs/msg/bool.hpp>
#include <std_msgs/msg/float32.hpp>
#include <vector>
#include <cmath>
#include <fstream>
#include <chrono>

struct Survivor {
  int id;
  double x,y,z;
  double time;
};

class MissionManager : public rclcpp::Node {
public:
  MissionManager() : Node("mission_manager") {
    this->declare_parameter("arena_size", 15.0);
    this->declare_parameter("max_mission_time", 1800.0);
    this->declare_parameter("max_survivors", 6);
    this->declare_parameter("home_x", 0.0);
    this->declare_parameter("home_y", 0.0);

    arena_=this->get_parameter("arena_size").as_double();
    max_time_=this->get_parameter("max_mission_time").as_double();
    max_surv_=this->get_parameter("max_survivors").as_int();
    home_x_=this->get_parameter("home_x").as_double();
    home_y_=this->get_parameter("home_y").as_double();

    map_sub_=this->create_subscription<nav_msgs::msg::OccupancyGrid>("/map", 10, std::bind(&MissionManager::mapCb, this, std::placeholders::_1));
    pose_sub_=this->create_subscription<geometry_msgs::msg::PoseStamped>("/offboard/pose", 10, std::bind(&MissionManager::poseCb, this, std::placeholders::_1));
    surv_sub_=this->create_subscription<geometry_msgs::msg::PoseStamped>("/person_detections", 10, std::bind(&MissionManager::survCb, this, std::placeholders::_1));
    e_stop_sub_=this->create_subscription<std_msgs::msg::Bool>("/emergency_stop", 10, std::bind(&MissionManager::eStopCb, this, std::placeholders::_1));

    state_pub_=this->create_publisher<std_msgs::msg::String>("/mission/state", 10);
    time_pub_=this->create_publisher<std_msgs::msg::Float32>("/mission/time_elapsed", 10);
    surv_list_pub_=this->create_publisher<std_msgs::msg::String>("/mission/survivors", 10);
    goal_pub_=this->create_publisher<geometry_msgs::msg::PoseStamped>("/exploration/goal", 10);
    abort_pub_=this->create_publisher<std_msgs::msg::Bool>("/mission/abort", 10);

    state_="IDLE";
    start_time_=this->now();
    timer_=this->create_wall_timer(std::chrono::seconds(1), std::bind(&MissionManager::fsm, this));
    RCLCPP_INFO(this->get_logger(), "C++ AirMouse Mission Manager arena %.1f max %.0fs", arena_, max_time_);
  }

private:
  void mapCb(const nav_msgs::msg::OccupancyGrid::SharedPtr msg) {
    int unknown=0;
    for (auto v: msg->data) if (v==-1) ++unknown;
    explored_=1.0 - (double)unknown/msg->data.size();
  }
  void poseCb(const geometry_msgs::msg::PoseStamped::SharedPtr msg) { pose_=*msg; has_pose_=true; }
  void survCb(const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
    // Check duplicate 1m
    for (auto &s: survivors_) {
      if (std::hypot(s.x-msg->pose.position.x, s.y-msg->pose.position.y)<1.0) return;
    }
    if ((int)survivors_.size()>=max_surv_) return;
    Survivor sv;
    sv.id=survivors_.size()+1;
    sv.x=msg->pose.position.x; sv.y=msg->pose.position.y; sv.z=msg->pose.position.z;
    sv.time=(this->now()-start_time_).seconds();
    survivors_.push_back(sv);
    RCLCPP_INFO(this->get_logger(), "Survivor %d at %.2f,%.2f total %zu/%d", sv.id, sv.x, sv.y, survivors_.size(), max_surv_);
    // Save json
    std::ofstream f("/tmp/nidar_survivors.json");
    f<<"[\n";
    for (size_t i=0;i<survivors_.size();++i) {
      f<<"  {\"id\":"<<survivors_[i].id<<",\"x\":"<<survivors_[i].x<<",\"y\":"<<survivors_[i].y<<"}";
      if (i+1<survivors_.size()) f<<",";
      f<<"\n";
    }
    f<<"]\n";
  }
  void eStopCb(const std_msgs::msg::Bool::SharedPtr msg) { if (msg->data) { emergency_=true; state_="ABORT"; } }

  void fsm() {
    std_msgs::msg::String s; s.data=state_; state_pub_->publish(s);
    double elapsed=(this->now()-start_time_).seconds();
    std_msgs::msg::Float32 t; t.data=elapsed; time_pub_->publish(t);
    if (elapsed>max_time_ && state_!="RETURN_HOME" && state_!="LAND" && state_!="COMPLETE") { state_="RETURN_HOME"; }

    if (state_=="IDLE") { state_="SETUP"; setup_time_=this->now(); }
    else if (state_=="SETUP") { if ((this->now()-setup_time_).seconds()>5.0) state_="ARM"; }
    else if (state_=="ARM") { state_="TAKEOFF"; takeoff_time_=this->now(); start_time_=this->now(); }
    else if (state_=="TAKEOFF") { if ((this->now()-takeoff_time_).seconds()>5.0) state_="EXPLORE_MAZE"; }
    else if (state_=="EXPLORE_MAZE") {
      if ((int)survivors_.size()>=max_surv_) state_="GENERATE_MAP";
      else if (explored_>0.90) state_="GENERATE_MAP";
    }
    else if (state_=="GENERATE_MAP") { state_="RETURN_HOME"; }
    else if (state_=="RETURN_HOME") {
      geometry_msgs::msg::PoseStamped home;
      home.header.frame_id="map"; home.header.stamp=this->now();
      home.pose.position.x=home_x_; home.pose.position.y=home_y_; home.pose.position.z=1.5; home.pose.orientation.w=1.0;
      goal_pub_->publish(home);
      if (has_pose_) {
        double dx=pose_.pose.position.x-home_x_, dy=pose_.pose.position.y-home_y_;
        if (std::hypot(dx,dy)<0.5) { state_="LAND"; saveReport(); }
      }
    }
    else if (state_=="LAND") { state_="COMPLETE"; }
    else if (state_=="ABORT") {
      std_msgs::msg::Bool b; b.data=true; abort_pub_->publish(b);
      state_="LAND";
    }
  }

  void saveReport() {
    std::ofstream f("/tmp/nidar_final_report.json");
    f<<"{\n";
    f<<"  \"mission\": \"NIDAR AirMouse C++\",\n";
    f<<"  \"arena_size\": "<<arena_<<",\n";
    f<<"  \"survivors_found\": "<<survivors_.size()<<",\n";
    f<<"  \"time\": "<<(this->now()-start_time_).seconds()<<"\n";
    f<<"}\n";
    RCLCPP_INFO(this->get_logger(), "Final report saved");
  }

  std::string state_;
  double arena_, max_time_, home_x_, home_y_, explored_=0;
  int max_surv_;
  geometry_msgs::msg::PoseStamped pose_;
  bool has_pose_=false, emergency_=false;
  std::vector<Survivor> survivors_;
  rclcpp::Time start_time_, setup_time_, takeoff_time_;
  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_sub_;
  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_sub_, surv_sub_;
  rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr e_stop_sub_;
  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr state_pub_, surv_list_pub_;
  rclcpp::Publisher<std_msgs::msg::Float32>::SharedPtr time_pub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr goal_pub_;
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr abort_pub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MissionManager>());
  rclcpp::shutdown();
  return 0;
}

File: src/air_mouse_mission/air_mouse_mission/mission_manager.py
#!/usr/bin/env python3
"""
mission_manager.py - NIDAR 2026 AirMouse Mission Manager
Mission Brief (from nidar.org.in/missions):
- Arena 15x15m, corridor width >=1m, height 8ft, room 2x2m
- Up to 6 survivors, entry/exit same (2ft x 2ft launch pad)
- Must generate 2D map, tag survivor locations, autonomous, no manual waypoint
- Max mission 30 mins, setup 5 mins
- Emergency stop / mission abort + failsafe required
- Scoring: map quality, survivors detected, time, return to home

This node is the high-level mission controller for AirMouse.
States:
IDLE -> SETUP -> ARM -> TAKEOFF -> EXPLORE_MAZE -> SEARCH_SURVIVORS -> GENERATE_MAP -> RETURN_HOME -> LAND -> COMPLETE
Handles failsafe, emergency stop, time monitoring.

Interfaces:
- Sub: /map, /person_detections, /offboard/pose, /emergency_stop
- Pub: /exploration/goal, /mission/state, /mission/time, /final_2d_map, /survivor_list
"""
import rclpy
from rclpy.node import Node
from nav_msgs.msg import OccupancyGrid
from geometry_msgs.msg import PoseStamped
from std_msgs.msg import String, Bool, Float32, Int32
import json
import time
import math
import os

class AirMouseMissionManager(Node):
    def __init__(self):
        super().__init__('air_mouse_mission_manager')
        self.declare_parameter('arena_size', 15.0)
        self.declare_parameter('max_mission_time', 1800.0)  # 30 mins
        self.declare_parameter('launch_pad_size', 0.6096)  # 2ft = 0.6096m
        self.declare_parameter('max_survivors', 6)
        self.declare_parameter('home_x', 0.0)
        self.declare_parameter('home_y', 0.0)

        self.arena_size = self.get_parameter('arena_size').value
        self.max_time = self.get_parameter('max_mission_time').value
        self.max_survivors = self.get_parameter('max_survivors').value
        self.home_x = self.get_parameter('home_x').value
        self.home_y = self.get_parameter('home_y').value

        # Mission state
        self.state = "IDLE"
        self.start_time = None
        self.survivors_found = []  # list of dicts
        self.map = None
        self.current_pose = None
        self.emergency_stop = False
        self.mission_complete = False

        # Subs
        self.map_sub = self.create_subscription(OccupancyGrid, '/map', self.map_callback, 10)
        self.pose_sub = self.create_subscription(PoseStamped, '/offboard/pose', self.pose_callback, 10)
        self.survivor_sub = self.create_subscription(PoseStamped, '/person_detections', self.survivor_callback, 10)
        self.e_stop_sub = self.create_subscription(Bool, '/emergency_stop', self.e_stop_callback, 10)
        self.map_3d_sub = self.create_subscription(OccupancyGrid, '/scan_3d_world', self.map_3d_callback, 10)

        # Pubs
        self.state_pub = self.create_publisher(String, '/mission/state', 10)
        self.time_pub = self.create_publisher(Float32, '/mission/time_elapsed', 10)
        self.survivor_list_pub = self.create_publisher(String, '/mission/survivors', 10)
        self.return_pub = self.create_publisher(PoseStamped, '/exploration/goal', 10)
        self.abort_pub = self.create_publisher(Bool, '/mission/abort', 10)

        # Timer 1Hz FSM
        self.timer = self.create_timer(1.0, self.fsm_loop)

        self.get_logger().info(f"AirMouse Mission Manager: Arena {self.arena_size}m, Max {self.max_time}s, Home ({self.home_x},{self.home_y})")

    def map_callback(self, msg):
        self.map = msg

    def map_3d_callback(self, msg):
        pass

    def pose_callback(self, msg):
        self.current_pose = msg

    def survivor_callback(self, msg):
        # Check if new survivor (distance >1m from existing)
        new = True
        for s in self.survivors_found:
            dx = s['x'] - msg.pose.position.x
            dy = s['y'] - msg.pose.position.y
            if math.hypot(dx, dy) < 1.0:
                new = False
                break
        if new and len(self.survivors_found) < self.max_survivors:
            entry = {
                'id': len(self.survivors_found)+1,
                'x': msg.pose.position.x,
                'y': msg.pose.position.y,
                'z': msg.pose.position.z,
                'time': time.time() - self.start_time if self.start_time else 0
            }
            self.survivors_found.append(entry)
            self.get_logger().info(f"Survivor {entry['id']} found at {entry['x']:.2f}, {entry['y']:.2f} - Total {len(self.survivors_found)}/{self.max_survivors}")

            # Publish list
            self.publish_survivors()

    def e_stop_callback(self, msg):
        self.emergency_stop = msg.data
        if self.emergency_stop:
            self.get_logger().warn("EMERGENCY STOP TRIGGERED!")
            self.state = "ABORT"

    def publish_survivors(self):
        msg = String()
        msg.data = json.dumps(self.survivors_found)
        self.survivor_list_pub.publish(msg)

        # Save to file for NIDAR scoring
        with open('/tmp/nidar_survivors.json', 'w') as f:
            json.dump(self.survivors_found, f, indent=2)

    def fsm_loop(self):
        # Publish state and time
        state_msg = String()
        state_msg.data = self.state
        self.state_pub.publish(state_msg)

        if self.start_time is not None:
            elapsed = time.time() - self.start_time
            time_msg = Float32()
            time_msg.data = elapsed
            self.time_pub.publish(time_msg)

            # Check max time
            if elapsed > self.max_time:
                self.get_logger().warn(f"Max mission time {self.max_time}s exceeded, returning home")
                self.state = "RETURN_HOME"

        # FSM
        if self.state == "IDLE":
            self.get_logger().info("IDLE -> SETUP (waiting for arm command)")
            self.state = "SETUP"
            self.setup_time = time.time()

        elif self.state == "SETUP":
            # 5 min setup time allowed, we simulate 5 sec
            if time.time() - self.setup_time > 5.0:
                self.get_logger().info("SETUP complete -> ARM")
                self.state = "ARM"

        elif self.state == "ARM":
            self.get_logger().info("ARM -> TAKEOFF")
            self.state = "TAKEOFF"
            self.takeoff_time = time.time()
            self.start_time = time.time()

        elif self.state == "TAKEOFF":
            if time.time() - self.takeoff_time > 5.0:
                self.get_logger().info("TAKEOFF complete -> EXPLORE_MAZE")
                self.state = "EXPLORE_MAZE"
                self.explore_start = time.time()

        elif self.state == "EXPLORE_MAZE":
            # Exploration managed by fuel_ros2_exploration
            # Check if exploration complete or all survivors found
            if len(self.survivors_found) >= self.max_survivors:
                self.get_logger().info(f"All {self.max_survivors} survivors found! -> GENERATE_MAP")
                self.state = "GENERATE_MAP"
            elif self.map is not None:
                # Check explored ratio
                unknown = sum(1 for v in self.map.data if v == -1)
                total = len(self.map.data)
                explored = 1 - unknown/total if total>0 else 0
                if explored > 0.90:
                    self.get_logger().info(f"Explored {explored*100:.1f}% -> GENERATE_MAP")
                    self.state = "GENERATE_MAP"

        elif self.state == "GENERATE_MAP":
            self.get_logger().info("Generating final 2D map for NIDAR scoring...")
            # Trigger map save via service or node
            self.save_final_map()
            self.state = "RETURN_HOME"

        elif self.state == "RETURN_HOME":
            # Publish home goal
            home_goal = PoseStamped()
            home_goal.header.frame_id = "map"
            home_goal.header.stamp = self.get_clock().now().to_msg()
            home_goal.pose.position.x = self.home_x
            home_goal.pose.position.y = self.home_y
            home_goal.pose.position.z = 1.5
            home_goal.pose.orientation.w = 1.0
            self.return_pub.publish(home_goal)

            if self.current_pose:
                dx = self.current_pose.pose.position.x - self.home_x
                dy = self.current_pose.pose.position.y - self.home_y
                if math.hypot(dx, dy) < 0.5:
                    self.get_logger().info("Returned to launch pad -> LAND")
                    self.state = "LAND"

        elif self.state == "LAND":
            self.get_logger().info("Landing on launch pad (2ft x 2ft)")
            self.state = "COMPLETE"
            self.mission_complete = True
            self.save_final_report()

        elif self.state == "COMPLETE":
            self.get_logger().info(f"MISSION COMPLETE! Survivors {len(self.survivors_found)}/{self.max_survivors}, Time {time.time()-self.start_time:.1f}s")

        elif self.state == "ABORT":
            self.get_logger().error("MISSION ABORTED - Emergency stop or failsafe")
            abort_msg = Bool()
            abort_msg.data = True
            self.abort_pub.publish(abort_msg)
            # Trigger landing
            self.state = "LAND"

    def save_final_map(self):
        # Save map as image and yaml for NIDAR
        if self.map is None:
            return
        # The actual saving is done by maze_mapping/map_saver, we just ensure files exist
        self.get_logger().info("Final map should be at /tmp/maze_map_with_persons.png and /tmp/maze_map.pgm")

    def save_final_report(self):
        report = {
            'mission': 'NIDAR 2026 AirMouse',
            'arena_size': self.arena_size,
            'max_time': self.max_time,
            'actual_time': time.time() - self.start_time if self.start_time else 0,
            'survivors_found': len(self.survivors_found),
            'survivors': self.survivors_found,
            'home': {'x': self.home_x, 'y': self.home_y},
            'complete': self.mission_complete
        }
        with open('/tmp/nidar_final_report.json', 'w') as f:
            json.dump(report, f, indent=2)
        self.get_logger().info(f"Final report saved to /tmp/nidar_final_report.json: {report}")

def main(args=None):
    rclpy.init(args=args)
    node = AirMouseMissionManager()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
